### PR TITLE
Update tab rendering logic and interaction

### LINKED ISSUE #8196: Performance: Change tab panels to only render current tab + manual activation for a11y
## Feature Description

This changes the tab implementation from "selection follows focus" with all panels preloaded in the DOM to "manual activation" as per WAI-ARIA guidelines.

This allows us to lazy-load initially inactive tabs + reduces initial render. It also improves keyboard navigation, as you can now switch from tab 1 to tab 4 without waiting for intermediate tabs to render. Instead, it requires the user to manually activate the fourth tab after moving focus there.

## Alternatives Considered

Current implementation is the primary alternative, but it's already being circumvented in some places.

## Additional Context

* [WAI-ARIA Authoring Practices 1.1 §6.4](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_selection_follows_focus)
* [_Example of Tabs with Manual Activation_](https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-2/tabs.html) from WAI-ARIA Authoring Practices demo implementations.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by barklund**
## Context

As per #8196, this does two things:

* Change the keyboard interaction for tab views to use manual interaction
* Renders only the current tab panel for tabbed displays

## User-facing changes

Only in the keyboard interaction mode for tab views.

### QA

This PR can be tested by following these steps:

1. Click a tab in the library tab list
2. Use arrow keys and/or home/end to move focus to another tab
3. Observe that only tab focus changes, not the current tab
4. With another tab than the current tab focused, click space/enter
5. Observe that the current tab changes

### UAT

- [x] UAT should use the same steps as above.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8196

### CODE REVIEW

File: assets/src/edit-story/components/library/libraryProvider.js [hunk 1]
```diff
@@ -158,22 +143,17 @@ function LibraryProvider({ children }) {
...
-
-    // if text pane hasn't been rendered until now fetch dynamically imported text sets
-    if (tab === TEXT.id && !previouslyRenderedTabs[tab]) {
+    async function getTextSets() {
+      const trackTiming = getTimeTracker('load_text_sets');
+      setTextSets(await loadTextSets());
+      trackTiming();
+    }
+    // if text sets have not been loaded but are needed fetch dynamically imported text sets
+    if (tab === TEXT.id && !Object.keys(textSets).length) {
```
Comment (miina) [inline assets/src/edit-story/components/library/libraryProvider.js]: This could also be moved directly to the text tab now instead perhaps?

File: assets/src/edit-story/components/tabview/index.js [hunk 2]
```diff
@@ -228,8 +237,13 @@ function TabView({
   useKeyDownEffect(ref, ['up', 'down'], () => {}, []);
 
   // Home/end keys navigate to first/last tab regardless of direction
-  useKeyDownEffect(ref, 'home', () => selectTabByIndex(0), [selectTabByIndex]);
-  useKeyDownEffect(ref, 'end', () => selectTabByIndex(-1), [selectTabByIndex]);
+  useKeyDownEffect(ref, 'home', () => focusTabByIndex(0), [focusTabByIndex]);
+  useKeyDownEffect(ref, 'end', () => focusTabByIndex(-1), [focusTabByIndex]);
+
+  // Enter/space to select currently focused tab.
+  useKeyDownEffect(ref, ['enter', 'space'], () => selectTab(focused.current), [
```
Comment (miina) [inline assets/src/edit-story/components/tabview/index.js]: Would this interfere with using space/enter to use the `QuickAction` `+` for adding a text element directly?

**Comment by miina** [comment]
@barklund Some tests need updating now.

File: assets/src/edit-story/components/tabview/index.js (hunk 2 above)
Comment (barklund) [reply assets/src/edit-story/components/tabview/index.js]: It did interfere - well spotted! I [wrapped it in a check](https://github.com/google/web-stories-wp/pull/8335/commits/1baf3b3ff6e3abc30223cca478314fcf05063908#diff-d0148c624c999bce1a708f7517bc0393d21e3828c4e696a8d8986dff4c0ddbc3R244-R255) for whether the keyboard focus is on a tab or not (and allowing default if not), and that seems to work great!

File: assets/src/edit-story/components/library/libraryProvider.js (hunk 1 above)
Comment (barklund) [reply assets/src/edit-story/components/library/libraryProvider.js]: Yeah, I wasn't sure why it was here in the first place, as it was always invoked the first time the tab was shown anyway, but I'm guessing there was a reason?

Comment (miina) [reply assets/src/edit-story/components/library/libraryProvider.js]: Hm, not sure what could be the reason, perhaps access to the `renderedTabs`? Looks like currently we can move it though, doesn't seem to have a reason to stay here anymore (as far as I can see 🤷 )

File: assets/src/edit-story/components/tabview/index.js (hunk 2 above)
Comment (miina) [reply assets/src/edit-story/components/tabview/index.js]: Yep, works well now 🎉

**Review by miina — APPROVED** [review]
🧶

**Review by merapi — APPROVED** [review]
LG, the only question I have is should we do the same for 3rd party media type tabs?

https://user-images.githubusercontent.com/21036224/126182129-cfd26f2d-fc3e-4c05-b3be-235e4684840e.mov



Side question about the second part of the video: is focus buggy or should it be like this? in the first pass it focuses on the first element and later on the currently selected element.

**Comment by barklund** [comment]
> LG, the only question I have is should we do the same for 3rd party media type tabs?

Oh, forgot about those tabs! I will port to this section too!

>  Screen.Recording.2021-07-19.at.16.57.37.mov 
> Side question about the second part of the video: is focus buggy or should it be like this? in the first pass it focuses on the first element and later on the currently selected element.

That's a bug, yes. I will fix that too - well spotted!

**Comment by swissspidy** [comment]
Looks like there's a valid e2e test failure related to inserting 3P Media

https://github.com/google/web-stories-wp/runs/3192962166#step:12:25

**Comment by barklund** [comment]
> Looks like there's a valid e2e test failure related to inserting 3P Media

Given that [the same e2e test clicks on elements in the other tabs in the same way](https://github.com/google/web-stories-wp/blob/add/8196-manual-tab-activation/packages/e2e-tests/src/specs/editor/media/insert3PMedia.js) and these don't fail, I don't actually think it's a stable bug. Will try re-running and see if it works then - I can't see why the GIF tab elements should be different as they all use the exact same code.

**Comment by swissspidy** [comment]
Hmm then perhaps it's related to the issues uncovered in #8320

https://github.com/google/web-stories-wp/pull/8320/files#diff-9cda36ed4a801ea560ad80cd5f191205d3578f2c0eb271504841d4d0f2ba09e1R64-R66

**Comment by barklund** [comment]
> Hmm then perhaps it's related to the issues uncovered in #8320

In [the next run](https://github.com/google/web-stories-wp/pull/8335/checks?check_run_id=3200203955#step:13:21), both Unsplash and Tenor fail - so it's "just" a very unstable test.

Would you be okay with merging here and filing a follow-up bug to find a proper fix?

**Comment by swissspidy** [comment]
The `should show that there is no poster attached to the story` pre-publish checklist test fails too.

Basically every test inserting media items is flakey.

I'd be fine with merging, but only with flakey tests disabled and the follow-up ticket getting P0

**Comment by barklund** [comment]
> I'd be fine with merging, but only with flakey tests disabled and the follow-up ticket getting P0

Oh, we already have [that ticket](https://github.com/google/web-stories-wp/issues/7481) as [one of those tests](https://github.com/google/web-stories-wp/blob/add/8196-manual-tab-activation/packages/e2e-tests/src/specs/editor/media/insert3PMedia.js#L51-L53) is already skipped! 😂 

I'll skip the other tests for the same reason and up-prioritize that ticket!

**Comment by swissspidy** [comment]
Well that ticket was created before the dynamic imports PR was created, so entirely different reasons there.
Media inserttion only working on second click is worrysome IMO

**Comment by miina** [comment]
> Media inserttion only working on second click is worrysome IMO

FYI in my local I'm not able to drag/insert media on the first click on `main` either. Tests on #8285  are also failing due to this, so the issue doesn't seem to be related to this PR (in case it's assumed it is).

**Comment by miina** [comment]
E.g. this happens when trying to drag media after reloading the page (on `main`, local env):
dragging

The [failing Karma test](https://github.com/google/web-stories-wp/pull/8285/checks?check_run_id=3218351018) has the same behaviour on #8285.

**Comment by swissspidy** [comment]
> > Media inserttion only working on second click is worrysome IMO
> 
> FYI in my local I'm not able to drag/insert media on the first click on `main` either. Tests on #8285 are also failing due to this, so the issue doesn't seem to be related to this PR (in case it's assumed it is).

This was introduced in #8320. See https://github.com/google/web-stories-wp/pull/8320#issuecomment-883629065 and 661451550cb06dfbb87c1b7b5e379ee29425275c

**Comment by swissspidy** [comment]
@googlebot I consent

**Comment by miina** [comment]
@barklund There seem to be 2 3rd party Media library icons currently, this seems to cause at least some of the test failures:
media

**Comment by miina** [comment]
@swissspidy I'm seeing this test failing here and [elsewhere, too,](https://github.com/google/web-stories-wp/pull/8638/checks?check_run_id=3309596029#step:7:8853) is this known to be a flakey test / have you noticed anywhere else? Otherwise might be good to skip this and create an issue for it:
`should show cards that require the `hasUploadMediaAction` permission`

**Comment by swissspidy** [comment]
> @swissspidy I'm seeing this test failing here and [elsewhere, too,](https://github.com/google/web-stories-wp/pull/8638/checks?check_run_id=3309596029#step:7:8853) is this known to be a flakey test / have you noticed anywhere else? Otherwise might be good to skip this and create an issue for it:
> `should show cards that require the `hasUploadMediaAction` permission`

@miina That should have been fixed yesterday in #8654. Make sure you merge latest changes from `main`.

File: packages/story-editor/src/components/library/libraryProvider.js [hunk 3]
```diff
@@ -105,25 +95,14 @@ function LibraryProvider({ children }) {
     () =>
       [
         MEDIA,
+        MEDIA3P,
         showMedia3p && MEDIA3P,
```
Comment (swissspidy) [inline packages/story-editor/src/components/library/libraryProvider.js]:
```suggestion
        showMedia3p && MEDIA3P,
```

**Comment by swissspidy** [comment]
> @barklund There seem to be 2 3rd party Media library icons currently, this seems to cause at least some of the test failures:
> media

Looks like a merge conflict issue, https://github.com/google/web-stories-wp/pull/8335/files#r687565939 should fix this

**Comment by miina** [comment]
@googlebot I consent.

**Comment by miina** [comment]
@barklund The Karma tests passed now, however, there's one oddity in Percy -- left you a comment. Most likely due to dynamic loading of the color picker library but PTAL just in case.

**Comment by barklund** [comment]
> @barklund The Karma tests passed now, however, there's one oddity in Percy -- left you a comment. Most likely due to dynamic loading of the color picker library but PTAL just in case.

That screenshot test will be changed when the new basic color picker is merged anyway, which is happening soon, so I don't think we need to deal with this right now.

File: packages/story-editor/src/components/library/libraryProvider.js [hunk 4]
```diff
@@ -95,7 +95,6 @@ function LibraryProvider({ children }) {
     () =>
       [
         MEDIA,
-        MEDIA3P,
```
Comment (barklund) [inline packages/story-editor/src/components/library/libraryProvider.js]: 🙄

File: packages/story-editor/src/components/canvas/karma/cloneSelection.karma.js [hunk 5]
```diff
@@ -94,12 +94,11 @@ describe('Clone element integration', () => {
   });
 
   it('should correctly clone 1 element', async () => {
-    // Select img1
-    await clickElement(img1.id);
+    // Select img2
+    await clickElement(img2.id);
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/karma/cloneSelection.karma.js:98]: Hm, this doesn't seem related to this PR though, but if it fixes things... 🤷

Comment (miina) [reply packages/story-editor/src/components/canvas/karma/cloneSelection.karma.js:98]: The main change should in theory be in the `clickElement`, changing the clicking location (the element didn't get selected properly since the selection frame was a bit shifted in the tests -- not sure why it stopped working now in this PR, I think this shifted selection frame issue we've had previously already in the tests, IIRC)
